nautobot_netbox_importer/diffsync/netbox.py

```diff
@@ -0,0 +1,497 @@
+    wrapper.add_field("site", None)
+
+    location_wrapper = wrapper.adapter.get_or_create_wrapper("dcim.location")
+    site_wrapper = wrapper.adapter.get_or_create_wrapper("dcim.site")
+    region_wrapper = wrapper.adapter.get_or_create_wrapper("dcim.region")
+
+    def importer(source: RecordData, target: RecordData) -> None:
+        parent = source.get("parent", None)
+        location = source.get("location", None)
+        site = source.get("site", None)
+        region = source.get("region", None)
+
+        if field.nautobot_name == "parent_id" and parent:
+            result = wrapper.get_pk_from_uid(parent)
+            wrapper.add_reference(result, wrapper)
+        elif location:
+            result = location_wrapper.get_pk_from_uid(location)
+            location_wrapper.add_reference(result, wrapper)
+        elif site:
+            result = site_wrapper.get_pk_from_uid(site)
+            site_wrapper.add_reference(result, wrapper)
+        elif region:
+            result = region_wrapper.get_pk_from_uid(region)
+            region_wrapper.add_reference(result, wrapper)
+        else:
+            result = None
+
+        target[field.nautobot_name] = result
+
+    field.set_importer(importer)
+
+
+def _define_units(field: SourceField) -> None:
+    field.set_nautobot_field(field.name)
+
+    def importer(source: RecordData, target: RecordData) -> None:
+        # NetBox 3.4 units is `list[int]`, previous versions are JSON string with list of strings
+        units = source.get(field.name, None)
+        if units in EMPTY_VALUES:
+            return
+
+        if isinstance(units, str):
+            units = json.loads(units)
+            if units:
+                units = [int(unit) for unit in units]
+
+        target[field.name] = units
+
+    field.set_importer(importer)
+
+
+def _role_definition_factory(adapter: SourceAdapter, role_content_type: str):
+    if role_content_type in adapter.wrappers:
+        role_wrapper = adapter.wrappers[role_content_type]
+    else:
+        role_wrapper = adapter.configure_model(
+            role_content_type,
+            nautobot_content_type="extras.role",
+            fields={"color": "color", "content_types": "content_types"},
+        )
+
+    def definition(field: SourceField) -> None:
+        field.set_role_importer(role_wrapper)
+
+    return definition
+
+
+def _setup_source() -> SourceAdapter:
+    """Setup NetBox models importers."""
+    adapter = SourceAdapter(name="NetBox")
+    adapter.configure(
+        ignore_fields=(
+            # Following fields fails for dcim.rack, see `./source.py` `_FORCE_FIELDS`
+            "created",
+            "last_updated",
+        ),
+        ignore_content_types={
+            "contenttypes.contenttype": None,  # Nautobot has own content types
+            "sessions.session": None,  # Nautobot has own sessions
+            "dcim.cablepath": None,  # Recreated in Nautobot
+            "admin.logentry": None,  # NetBox 3.0 TBD verify
```

Comment:
what's the strategy to handle different Netbox versions?

Reply:
For now, we support 3.0 - 3.4, adding 3.5 and 3.6. The importer is created dynamically based on the current import data and Nautobot version. Till now, it's sufficient enough. If it will not work correctly, we can add versions constraints to `configure_model`, so that configuration will be applied only for particular Nautobot / NetBox versions.